lib/index.js

```diff
@@ -0,0 +1,220 @@
+'use strict'
+
+const Promise = require('bluebird')
+const yaml = require('js-yaml')
+const Joi = require('joi')
+const hostname = require('@runnable/hostname')
+
+const removeNullAndUndefinedKeys = require('./util').removeNullAndUndefinedKeys
+const validate = require('./util').validate
+const Parser = require('./parser')
+const Mapper = require('./mapper')
+const Warning = require('./warning')
+
+const dockerComposeFileSchema = Joi.object({
+  version: Joi.string().valid('2').required(),
+  services: Joi.object().required()
+}).unknown().required()
+
+const arrayOfStrings = Joi.array().items(Joi.string())
+
+const serviceSchemaKeys = {
+  build: Joi.alternatives().try(Joi.string(), Joi.object()),
+  image: Joi.string(),
+  command: Joi.alternatives().try(Joi.string(), arrayOfStrings),
+  ports: arrayOfStrings,
+  expose: arrayOfStrings,
+  links: arrayOfStrings,
+  depends_on: arrayOfStrings,
+  environment: Joi.alternatives().try(Joi.object(), arrayOfStrings)
+}
+const serviceSchema = Joi.object(serviceSchemaKeys).unknown().required()
+
+module.exports = class DockerComposeParser {
+
+  /**
+   * Parse a docker-compose.yml file into an array of instances that can be
+   * used by Runnable API Client
+   *
+   * @param {Object} opts
+   * @param {String} opts.dockerComposeFileString - YML File string
+   * @param {String} opts.repositoryName - Name of repository in GH or other service
+   * @param {String} opts.userContentDomain - User content domain for API environment. Used for creating hostnames.
+   * @param {String} opts.ownerUsername - Organization name in Runnable. Used for creating hostnames.
+   * @return {Promise}
+   * @resolves {Array<Object>}
+   */
+  static parse (opts) {
+    const dockerComposeFileString = opts.dockerComposeFileString
+    const repositoryName = opts.repositoryName
+    const userContentDomain = opts.userContentDomain
+    const ownerUsername = opts.ownerUsername
+
+    return DockerComposeParser._parseDockerComposeFile(dockerComposeFileString)
+    .then(services => DockerComposeParser._convertToArrayAndAddMetadata(services))
+    .map(service => DockerComposeParser._serviceParser({ service, repositoryName }))
```

Comment:
nit: `_serviceParser` does not really tell me the intent of this function. Try using an `action` verb like `_parseService` otherwise to me it looks like you're creating an object called `serviceParser`